ansible_collections/arista/avd/plugins/plugin_utils/topology_generator_utils.py

```diff
@@ -0,0 +1,715 @@
+
+    if "groups" in data and data["groups"]:
+        for group in data["groups"]:
+            root = find_root_nodes(group, root)
+        return root
+
+
+def create_graph_dict(ol, inventory_group, nodes=None, node_neighbor_dict=None):
+    """
+    Create a dictionary of rank/levels and it's corresponding node list
+
+    Args:
+      ol: List of dictionaries with nested "name", "nodes" and "groups" keys
+      nodes: List of nodes
+      node_neighbor_dict: Dictionary with nodes and neighbors list
+                        ex- {'0': ['super-spine1', 'super-spine2']}
+    Returns:
+       nodes: List of nodes which are present in the graph
+       node_neighbor_dict: Dictionary of node and its corresponding neighbors
+    """
+    if nodes is None:
+        nodes = []
+    if node_neighbor_dict is None:
+        node_neighbor_dict = {}
+    for group in ol:
+        if "nodes" in list(group.keys()) and group["nodes"]:
+            for node in group["nodes"]:
+                if  (node["name"] in inventory_group) or (node["name"] == "0"):
+                    nodes.append(node["name"])
+                    neighbors = []
+                    for neighbor in node["neighbors"]:
+                        if  neighbor["neighborDevice"] in inventory_group:  
+                            node_detail_dict = {}
+
+                            node_detail_dict["nodePort"] = neighbor["port"].replace("Ethernet", "")
+
+                            node_detail_dict["neighborDevice"] = neighbor["neighborDevice"]
+
+                            node_detail_dict["neighborPort"] = neighbor["neighborPort"].replace("Ethernet", "")
+
+                            neighbors.append(node_detail_dict)   
+                    node_neighbor_dict[node["name"]] = neighbors
+        if "groups" in group and group["groups"]:
+            create_graph_dict(group["groups"], inventory_group, nodes, node_neighbor_dict)
+     
+    return nodes, node_neighbor_dict
+
+
+def find_node_levels(graph, start_node, node_list):
+
+    """
+    Function to determine level of each node starting from start_node using BFS algorithm
+    Args:
+        graph: dictionary with node and its neighbors list
+        start_node: Starting point/root node of the graph
+        node_list: List of nodes
+    Returns:
+         dict: Return a dictionary with level as key and node list as value
+    """
+
+    # array to store level of each node
+    node_level_dict = {}
+    marked = {node: False for node in node_list}
+
+    # create a queue
+    que = queue.Queue()
+    # enqueue element x
+    que.put(start_node)
+    # initialize level of start_node
+    # node to 0
+    node_level_dict[start_node] = 0
+    # marked it as visited
+    marked[start_node] = True
+    # do until queue is empty
+    while not que.empty():
+        # get the first element of queue
+        start_node = que.get()      
+        # traverse neighbors of node start_node
+        if start_node in graph:  #key checking
```

Comment:
formating should fail on the comment - should be (if inline comments are accepted):

```suggestion
        if start_node in graph:  # key checking
```